File: examples/experimental/agent_builder/customer_care_style/toolkits/banking_mcp_server/src/creditCard.ts
/**
 * Credit Card Tools Module
 *
 * Contains tool definitions for credit card operations
 */

import { z } from 'zod';
import { CreditCardService } from './creditCardService';

/**
 * Localized message templates for credit card balance display
 * Templates use {variable} placeholders that get replaced with actual values
 */
const messageTemplates: Record<string, string> = {
  'en-US': `Credit Card Balance:
Card: {cardNumber} ({cardType})
Current Balance: {currentBalance}
Available Credit: {availableCredit}
Credit Limit: {creditLimit}
Minimum Payment: {minimumPayment}
Payment Due: {paymentDueDate}`,

  'es-MX': `Saldo de Tarjeta de Crédito:
Tarjeta: {cardNumber} ({cardType})
Saldo Actual: {currentBalance}
Crédito Disponible: {availableCredit}
Límite de Crédito: {creditLimit}
Pago Mínimo: {minimumPayment}
Fecha de Pago: {paymentDueDate}`,

  'fr-FR': `Solde de Carte de Crédit:
Carte: {cardNumber} ({cardType})
Solde Actuel: {currentBalance}
Crédit Disponible: {availableCredit}
Limite de Crédit: {creditLimit}
Paiement Minimum: {minimumPayment}
Date d'Échéance: {paymentDueDate}`,

  'de-DE': `Kreditkartensaldo:
Karte: {cardNumber} ({cardType})
Aktueller Saldo: {currentBalance}
Verfügbares Guthaben: {availableCredit}
Kreditlimit: {creditLimit}
Mindestzahlung: {minimumPayment}
Fälligkeitsdatum: {paymentDueDate}`,
};

/**
 * Get credit card balance tool definition
 *
 * This tool demonstrates:
 * 1. Using JWT token from CONTEXT VARIABLES (customer-provided via API)
 * 2. Using locale from SYSTEM CONTEXT (provided by Watson Orchestrate)
 */
export const getCreditCardBalanceTool = {
  name: 'get_credit_card_balance',
  config: {
    title: 'Get Credit Card Balance',
    description:
      'Retrieve current credit card balance and details for the authenticated customer',
    inputSchema: {},
    // outputSchema: {
    //   cardNumber: z.string(),
    //   cardType: z.string(),
    //   currentBalance: z.number(),
    //   availableCredit: z.number(),
    //   creditLimit: z.number(),
    //   minimumPayment: z.number(),
    //   paymentDueDate: z.string(),
    //   lastStatementDate: z.string(),
    //   lastStatementBalance: z.number(),
    // },
  },
  handler: async (args: any, extra: any) => {
    // Extract JWT token from CONTEXT VARIABLES (passed by API caller)
    const jwtToken = extra?._meta?.['com.ibm.orchestrate/context']?.jwtToken;

    if (!jwtToken) {
      throw new Error(
        'JWT token is required. It must be provided in context variables by the API caller.',
      );
    }

    // Use JWT token to authenticate with backend service
    const cardBalance = CreditCardService.getCreditCardBalance(jwtToken);

    // Extract locale from SYSTEM CONTEXT with fallback to en-US
    const locale =
      extra?._meta?.['com.ibm.orchestrate/systemcontext']?.locale || 'en-US';

    // Get message template for locale, fallback to en-US if locale not supported
    const template = messageTemplates[locale] || messageTemplates['en-US'];

    // Fill in template with actual values
    const message = template
      .replace('{cardNumber}', cardBalance.cardNumber)
      .replace('{cardType}', cardBalance.cardType)
      .replace('{currentBalance}', `$${cardBalance.currentBalance.toFixed(2)}`)
      .replace(
        '{availableCredit}',
        `$${cardBalance.availableCredit.toFixed(2)}`,
      )
      .replace('{creditLimit}', `$${cardBalance.creditLimit.toFixed(2)}`)
      .replace('{minimumPayment}', `$${cardBalance.minimumPayment.toFixed(2)}`)
      .replace('{paymentDueDate}', cardBalance.paymentDueDate);

    return {
      content: [
        {
          type: 'text' as const,
          text: message,
          annotations: { audience: ['user'] },
        },
      ],
    };
  },
};

/**
 * Export all credit card tools
 */
export const creditCardTools = [getCreditCardBalanceTool];

File: examples/experimental/agent_builder/customer_care_style/toolkits/banking_mcp_server/py_server/src/credit_card.py
"""
Credit Card Tools Module

Contains tool definitions for credit card operations.
Uses dict-based tool definitions to match the TypeScript implementation.
"""

from .credit_card_service import CreditCardService


# Localized message templates for credit card balance display
# Templates use {variable} placeholders that get replaced with actual values
MESSAGE_TEMPLATES: dict[str, str] = {
    "en-US": """Credit Card Balance:
Card: {card_number} ({card_type})
Current Balance: {current_balance}
Available Credit: {available_credit}
Credit Limit: {credit_limit}
Minimum Payment: {minimum_payment}
Payment Due: {payment_due_date}""",
    "es-MX": """Saldo de Tarjeta de Crédito:
Tarjeta: {card_number} ({card_type})
Saldo Actual: {current_balance}
Crédito Disponible: {available_credit}
Límite de Crédito: {credit_limit}
Pago Mínimo: {minimum_payment}
Fecha de Pago: {payment_due_date}""",
    "fr-FR": """Solde de Carte de Crédit:
Carte: {card_number} ({card_type})
Solde Actuel: {current_balance}
Crédit Disponible: {available_credit}
Limite de Crédit: {credit_limit}
Paiement Minimum: {minimum_payment}
Date d'Échéance: {payment_due_date}""",
    "de-DE": """Kreditkartensaldo:
Karte: {card_number} ({card_type})
Aktueller Saldo: {current_balance}
Verfügbares Guthaben: {available_credit}
Kreditlimit: {credit_limit}
Mindestzahlung: {minimum_payment}
Fälligkeitsdatum: {payment_due_date}""",
}


async def get_credit_card_balance_handler(args: dict, extra: dict) -> dict:
    """
    Get credit card balance tool handler.

    This tool demonstrates:
    1. Using JWT token from CONTEXT VARIABLES (customer-provided via API)
    2. Using locale from SYSTEM CONTEXT (provided by Watson Orchestrate)
    """
    _meta = extra.get("_meta", {})

    # Extract JWT token from CONTEXT VARIABLES (passed by API caller)
    jwt_token = _meta.get("com.ibm.orchestrate/context", {}).get("jwtToken")

    if not jwt_token:
        raise ValueError(
            "JWT token is required. It must be provided in context variables by the API caller."
        )

    # Use JWT token to authenticate with backend service
    card_balance = CreditCardService.get_credit_card_balance(jwt_token)

    # Extract locale from SYSTEM CONTEXT with fallback to en-US
    locale = _meta.get("com.ibm.orchestrate/systemcontext", {}).get("locale", "en-US")

    # Get message template for locale, fallback to en-US if locale not supported
    template = MESSAGE_TEMPLATES.get(locale, MESSAGE_TEMPLATES["en-US"])

    # Fill in template with actual values
    message = (
        template.replace("{card_number}", card_balance.card_number)
        .replace("{card_type}", card_balance.card_type)
        .replace("{current_balance}", f"${card_balance.current_balance:.2f}")
        .replace("{available_credit}", f"${card_balance.available_credit:.2f}")
        .replace("{credit_limit}", f"${card_balance.credit_limit:.2f}")
        .replace("{minimum_payment}", f"${card_balance.minimum_payment:.2f}")
        .replace("{payment_due_date}", card_balance.payment_due_date)
    )

    return {
        "content": [
            {
                "type": "text",
                "text": message,
                "annotations": {"audience": ["user"]},
            }
        ]
    }


# Tool definitions in dict format (matching TypeScript structure)
get_credit_card_balance_tool = {
    "name": "get_credit_card_balance",
    "config": {
        "title": "Get Credit Card Balance",
        "description": (
            "Retrieve current credit card balance and details for the authenticated customer"
        ),
        "inputSchema": {
            "type": "object",
            "properties": {},
            "required": [],
        },
    },
    "handler": get_credit_card_balance_handler,
}

# Export all credit card tools
credit_card_tools = [get_credit_card_balance_tool]
